consensus/proto_array/src/proto_array.rs

```diff
@@ -990,6 +996,70 @@ impl ProtoArray {
             .unwrap_or(false)
     }
 
+    /// Returns `true` if `root` is equal to or a descendant of
+    /// `self.finalized_checkpoint`.
+    ///
+    /// Notably, this function is checking ancestory of the finalized
+    /// *checkpoint* not the finalized *block*.
+    pub fn is_finalized_checkpoint_or_descendant<E: EthSpec>(&self, root: Hash256) -> bool {
+        let finalized_root = self.finalized_checkpoint.root;
+        let finalized_slot = self
+            .finalized_checkpoint
+            .epoch
+            .start_slot(E::slots_per_epoch());
+
+        let mut node = if let Some(node) = self
+            .indices
+            .get(&root)
+            .and_then(|index| self.nodes.get(*index))
+        {
+            node
+        } else {
+            // An unknown root is not a finalized descendant. This line can only
+            // be reached if the user supplies a root that is not known to fork
+            // choice.
+            return false;
+        };
+
+        // The finalized and justified checkpoints represent a list of known
+        // ancestors of `node` that are likely to coincide with the store's
+        // finalized checkpoint.
+        //
+        // Don't continue checking these values for ancestors. If they don't
+        // match for the child then they're unlikely to start matching for its
+        // ancestors.
```

Comment:
This comment confused me a bit, is it leftover from a previous implementation that checked these checkpoints on the ancestors?

Reply:
It's more of a reasoning why these checks are outside the loop, rather than inside it. Happy to re-word if you can show me which part was most confusing :)

Reply:
Ohhhh I get it now, I took it without the context of the loop below

Reply:
Happy to leave as-is

Reply:
I tweaked the comment to avoid this confusion in the future:

```
// Run this check once, outside of the loop rather than inside the loop.
// If the conditions don't match for the child then they're unlikely to
// start matching for its ancestors.
```